### ISSUE TITLE
Setting the Token for AWSCognitoIdentityProvider.defaultCognitoIdentityProvider().describeUserPool

### ISSUE THREAD

**Issue by BruceBuckland** [awslabs/aws-sdk-ios-samples#108]
Attempting to integrate User Pools and Cognito Id's

How do I set the access token? I have tried everything, and it seems I am signed in, and I did manage to get the tokens from one or more calls, but I cannot SET them when trying to call describe userpool

I try this

```
        let descriptionRequest = AWSCognitoIdentityProviderDescribeUserPoolRequest()
        descriptionRequest.userPoolId = K.COGNITO_USER_POOL_ID
        let provider = AWSCognitoIdentityProvider.defaultCognitoIdentityProvider()

        provider.describeUserPool(descriptionRequest).continueWithBlock{ (task) in...
```

and I get the error

```
    Connection = "keep-alive";
    "Content-Length" = 89;
    "Content-Type" = "application/x-amz-json-1.1";
    Date = "Sat, 03 Sep 2016 21:26:39 GMT";
    "x-amzn-ErrorMessage" = "Missing Authentication Token";
    "x-amzn-ErrorType" = "MissingAuthenticationTokenException:";
    "x-amzn-RequestId" = "XXXXXXXXXXXXXXXX...c4ab";

```

So my question is two fold:

1) What token does it want?  - here are the ones I know about and have access to:
refreshToken, IdToken, AccessToken, SessionToken.  Which of those is the token that Cognito ID traded to me when I logged in... I suspect NONE of them are that token and I suspect that is the token I need.
2) How do I make the token part of the request.  I assume the IOS SDK for the API is supposed to be doing that, because the api makes it pretty hard to get the tokens in the first place.

I have tried the following, none of them resolve the issue.  I have determined that my User Pool knows I am logged in, that is it has an logins array with my userpool id and a token of some kind.  I of course cannot get cognito to tell me whether it knows about the fact that my user is logged in because the function to do that is deprecated and returns null, so I have no way of knowing that cognito thinks I am logged in.  if I do self.pool.currentUser().signedIn is is always true.  Is this Cognito ID telling me I am signed in, or is this User Pools?

Here is what I do after the user is logged in (all of this was my attempt to "wave a dead chicken" and make Cognito work with Cognito User Pools.  The most helpful documentation on this has been the BYOI (bring your own identity?) documentation because it seems that you have treated Cognito User pools as a developer identity system (but unfortunately I cannot plug in a SAML ID or a OAuth Id because neither seems to work (and the documentation doesn't say I need to.. but I must be forgetting something, or perhaps you have never tried Cognito User Pools with Cognito ID (if that is the case it sure would be nice to hear it))))

Here is what I am currently doing ... (I know a lot of this is unnecessary... consider it API archeology trying to figure out how to use Cognito with Cognito Identity

```
user = pool!.currentUser()
self.user!.getSession().continueWithSuccessBlock() { (task) in // starts initAuth if needed self.user!.getDetails().continueWithSuccessBlock() { (task) in ....
self.appDelegate.credentialsProvider!.getIdentityId().continueWithBlock { (task: AWSTask!) ....
self.appDelegate.credentialsProvider?.credentials().continueWithBlock() { (task: AWSTask!) -> AnyObject! in ....

// and then again... (because I wanted to try this AFTER i was logged in)
self.appDelegate.user?.getSession().continueWithBlock() { (task: AWSTask!) -> AnyObject! in....

// and then look at the pool logins ... and they show the login
 self.appDelegate.pool?.logins().continueWithBlock() { (task: AWSTask!) -> AnyObject! in

//At this point.. If I try assumeRoleWithWebIdentity i get Sender, Message: Not authorized to perform sts:AssumeRoleWithWebIdentity, Code: AccessDenied because it still thinks I am an unauthenticated user

// also at this point if I try 
provider.describeUserPool(descriptionRequest).continueWithBlock{ (task) in...
// I get the error described above (missing token).
```

I know you guys are in the middle of evolving cognito, but it would really be useful if you were to help us by explaining some of the roles you envision/have designed. The terminology is confusing., and not used consistently. For instance:

Cognito User Pools is an authentication provider? and IdP (like with SAML)... or what? Is it a SAML IdP or is it OAuth or neither? What is it, and how do I use it as a developer provided identity system or is there a better way to use it with cognito id?  Do I need to treat it like a developer provided IdP?

Cognito ID Pools apparently are NOT an IdP... instead id pools are some sort of middleman federating IdP's, yet when defining IAM roles, it is treated as the trusted entity, and when initializing the Cognito data structures it is passed as a credentials provider... what is that?

When I initialize a credentialsProvider I have to give it an identityProvider (I give it User Pools) but the credential provider is Cognito Identity... so shouldn't it be the identity provider?  Should I be initializing it with itself?

And does the order of initialization matter?  To make a serviceConfiguration you need a credentialsProvider but to make a credentials provider you need an identityProvider I.E. a user pool and to get a userPool you need a serviceConfiguration!  So some one of these objects needs to be instantiated with a null pointer to start with, and I chose serviceConfiguration and passed it an optional, but is that correct? And do I need to re-initialize or update anything after login?

I could find no example where AWS uses both Cognito User Pools AND Cognito together.

**Comment by helloniklas** [comment]
@BruceBuckland Totally agree with this, the APIs are a mess and hard to know how to use this properly.

I seem to be able to set the credentials provider but first sending in nil, then doing

```
        let pool = AWSCognitoIdentityUserPool(forKey: "UserPool")
         let credentialsProvider = AWSCognitoCredentialsProvider(regionType: .usEast1, identityPoolId: Constants.identityPoolId, identityProviderManager:pool)
credentialsProvider.getIdentityId().continue...
```

Which then does give me back the cognito id for that user.

However, when I then try and use the AWSDynamoDBObjectMapper for instance, it still comes back and says ""Request is missing Authentication Token" so I'm lost here as well.

**Comment by BruceBuckland** [comment]
@helloniklas Yesterday I did make some progress, but still have not got the ability to get myself to the Authenticated Role (I am stuck in unauthenticated).  I suspect that they don't really have this working.

And while this may seem like whining, they have chosen very poor names for things both from a marketing and programming point of view.  Start with Cognito Id Pools, then Cognito User Pools then figure out if you can which of classes apply to which!  And it helps when they refer to things like a "Cognito Id User" ... Ha!  

But I think the system would be great for my app, if I could just understand.

In my archeology I recently unearthed the following tidbits.

1) You can use IAM to create an identityProvider and CUPs  (Cognito User Pools) can fulfill that role. You create the identity Provider as a Open ID provider.  The url is http://cognito-idp.us-east-1.amazonaws.com/<userpool id>, and the audience is the user pool id (I think... but I get errors when trying to use the audience to authenticate regarding the audience.  I don't know if this is because it is not a valid audience, or because it cannot find the audience in my trust policy or access policy.  Any help there would be appreciated.  here is the error)
`<ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <Error>
    <Type>Sender</Type>
    <Code>InvalidIdentityToken</Code>
    <Message>Incorrect token audience</Message>
  </Error>
  <RequestId>35eb5cac-739f-11e6-85f3-e381d48bee15</RequestId>
</ErrorResponse>
`
2) I think what this means is that it is possible to use CUPs as an OpenID provider for developer authenticated identities (and the AWSCognitoIdentityProvider and AWSCognitoIdentityProviderManager are thus already implemented, by AWS.. they just call it user pools).  And the nice thing about this is that the AWS documentation for BYOI (Bring your own identity aka developer authenticated identities) is really not all that bad (because they have lots of people doing it.. so it has to work.  So my next try is to drop the cognito federation and just try to get BYOI working with CUPs as my BYOI.  They don't say it works, but CUPs seems to behave correctly as an OpenID provider.

3) I found that the way you initialize the credentialsProvider, which form of the call you use, alters the "authFlow" that the interface thinks you are going to use.  But I haven't figured out which init causes which authFlow, or if this is an error (an artifact) or part of the interface design.

Lastly you mention that getIdentityId() gets you a cognito id, which is cool.  And I do that and get an ID.... But how do you know that it is a cognito ID?  How do I recognize a cognito ID vs a token from CUPs?

Thanks for your support.  I really think if we can figure out how to make this stuff work and federate ID's with userPools and other providers, it would be great for my app.  

I am inspired by the product... I must say that I am underwhelmed by the documentation!

**Comment by helloniklas** [comment]
@BruceBuckland Thanks for the update. I've now gotten to the stage where it does auth, but it never assumes the auth role, only the unauth role, not sure why. I have used the BYIO in the past, and that worked, but for this project I thought User Pools would be a good fit.

The cognitoID I get back from getIdentityId() matches the one in the identity browser when I log into AWS console, so it's correct.

Someone must have go this working as it's little point to use these User Pools if you can't then access resources on AWS.

Totally the agree on the name choice, it's very poorly named and makes for a confusing API. To many different options to call, sometimes they need to be nil, sometimes not.

Please make a post if you get it all working.

**Comment by BruceBuckland** [comment]
@helloniklas Ooooh... Thank you.  I did see the Cognito ID in the identity browser, but those are permanent ID's. ( I thought the cognito ID was supposed to be temporary.  Or maybe there is some temporary credentials that go with it. )

**Comment by helloniklas** [comment]
@BruceBuckland the cognitoId stays the same for the same user I believe. Even if they log out and then in on another device. That's the way you can tie a specific user to an item in a dynamodb user table to store other stuff.

**Comment by BruceBuckland** [comment]
I should say that i did eventually make some progress on this.  

The key thing I learned was that if you go in the identity browser in the federated identity console, you can see if an identity has logins.  If it does, then at some point that identity was connected with a user (and his user pool)(and will be connected again with next time that user logs in).

I had some code that would not get a login, and then i Deleted the pool and recreated it, and it started working.  I think Federated Identities can get "wedged" in some way, but I have not been able to reproduce that.  

It works now as long as I:
1) create a service configuration (I did it with nil credentials provider)
2) create a user pool configuration
3) register the user pool configuration with the service configuration (convention says you use key "UserPool")
4) Instantiate the pool instance by key "UserPool" using AWSCognitoIdentityUserPool
5) Instantiate a credentials provider using AWSCognitoCredentialsProvider using your Federated Identities ID and setting the IdentityProviderManager as your pool instance. (note ... this approach is fine for authenticating with a single IdentityProvider (like in this case my User Pool) but if you want to federate different identities you need to build your own IdentityProviderManager to manage and accumulate all the logins from the different providers into a single list). But since I am only using user pools for login, and since the AWSCognitoIdentityUserPool can return logins (it conforms to IdentityProvider and IdentityProviderManager protocols) I can just use the pool as my IPM.
6) Not required, but at this point I set the AWSServiceManager defaultServiceConfiguration
7) set the pool delegate to whatever you want if you are using the delegate based authentication.

That setup is enough to work.   But you still won't get logins on your identity browser. For that you need to:

1) get authenticated using the userpool
2) after being authenticated do credentialsProvider.getIdentityId

at that point you are logged in and capable of doing things that require the use of the authenticated IAM role of your Identity Pool.

**Comment by ohsomofo** [comment]
@BruceBuckland Thank you for your summary. Would like to confirm that the above steps work for me as well. However I ran into another problem and would like to add one thing if anyone else happens on this thread after pulling most of their hair out. 

TLDR; Do not use the Info.plist to setup AWS services in an iOS application. Configure all your services in code. If you have used the Mobile Hub as a starting point you had instructions to copy a whole set of AWS configuration info into your Info.plist. That information can conflict with settings from code and should be removed.

It's probably a good idea to not have any reference to AWS setup in your info.plist. Do all of that in code. If you have everything configured properly in code but happen to have entries in the Info.plist, even if they are the same as what's in code, those entries may still cause issues when trying to access other AWS services.

For instance I had originally started with the Mobile Hub to setup an example app and setup my Info.plist accordingly. I ended up removing everything related to the Mobile Hub and starting from scratch but had left those keys in place and adjusted them as needed. When I finally got everything working for login with an authenticated user I was still running into an issue that whenever I tried to make a DynamoDB call I would get a new unauthenticated identity show up in the identity pool and the credentials provider would cache that.

After deleting the Info.plist entries under AWS for IdentityProvider and DynamoDB everything now works as expected. I have no idea why that may have been an issue but if you are able to create an authenticated identity in your identity pool but attempts to use other AWS services creates a new unauthenticated identity which gets cached then maybe it's the same thing I ran into.

**Comment by encio** [comment]
Hi I've been following the steps and here is I am so far.

Unfortunately I'm stuck in 

> 2) after being authenticated do credentialsProvider.getIdentityId

AppDelegate.swift 
```
func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {

//1) create a service configuration (I did it with nil credentials provider)
        let serviceConfiguration = AWSServiceConfiguration(region: MY.COGNITO_REGIONTYPE, credentialsProvider: nil)

//2) create a user pool configuration
        let userPoolConfiguration = AWSCognitoIdentityUserPoolConfiguration(clientId: MY.COGNITO_USER_POOL_APP_CLIENT_ID, clientSecret: MY.COGNITO_USER_POOL_APP_CLIENT_SECRET, poolId: MYCOGNITO_USER_POOL_ID)

//3) register the user pool configuration with the service configuration (convention says you use key "UserPool")       AWSCognitoIdentityUserPool.registerCognitoIdentityUserPoolWithConfiguration(serviceConfiguration, userPoolConfiguration: userPoolConfiguration, forKey: "UserPool")

//4) Instantiate the pool instance by key "UserPool" using AWSCognitoIdentityUserPool
        let pool = AWSCognitoIdentityUserPool(forKey: "UserPool")

// 5) Instantiate a credentials provider using AWSCognitoCredentialsProvider using your Federated Identities ID and setting the IdentityProviderManager as your pool instance.
        let credentialsProvider = AWSCognitoCredentialsProvider(regionType: .USEast1,
                                                            identityPoolId:"us-east-1:MY.IDENTITY_ID", identityProviderManager:pool)

//6) Not required, but at this point I set the AWSServiceManager defaultServiceConfiguration
        let defaultServiceConfiguration = AWSServiceConfiguration(region: MY.COGNITO_REGIONTYPE, credentialsProvider: credentialsProvider)
//IS THIS WHAT YOU MEANT?? 
        AWSServiceManager.defaultServiceManager().defaultServiceConfiguration = defaultServiceConfiguration

//7) set the pool delegate to whatever you want if you are using the delegate based authentication. 
        pool.delegate = self

  return true
 }

```

This is part where my user is authenticated via userpool
In my `LoginViewController`

    
```
func didCompletePasswordAuthenticationStepWithError(error: NSError?) {

//1) get authenticated using the userpool
//This part of the code is where I authenticate my user and if there is no error, the log in is succesful
            if error != nil {
                    // 2) after being authenticated do credentialsProvider.getIdentityId
                    // This is where Im stuck. I cant figure how to to it from here
            }
    }
```

Please help me how to do the last part

**Comment by BruceBuckland** [comment]
This may help conceptually, it is from my notes on Cognito [cognito diagram.pdf](https://github.com/awslabs/aws-sdk-ios-samples/files/542234/cognito.diagram.pdf)

What I did was this:
        
```
        serviceConfiguration = AWSServiceConfiguration(region:.USEast1, credentialsProvider:nil)
        
        // now create a user pool configuration with our keys and ids
        
        let userPoolConfiguration = AWSCognitoIdentityUserPoolConfiguration(clientId: K.COGNITO_USER_POOL_APP_CLIENT_ID, clientSecret: K.COGNITO_USER_POOL_APP_CLIENT_SECRET, poolId: K.COGNITO_USER_POOL_ID)
        
        // now we register that pool with the service configuration which will allow us
        // to use the pool
        
        AWSCognitoIdentityUserPool.registerCognitoIdentityUserPoolWithConfiguration(serviceConfiguration, userPoolConfiguration: userPoolConfiguration, forKey: "UserPool")
        

        
        // and this gets a working pool.
        
        self.pool = AWSCognitoIdentityUserPool(forKey: "UserPool")
        
        
        // this sets the credentials provider on the CFI pool and establishes the CUP
        // as the identityProviderManager
        
        credentialsProvider = AWSCognitoCredentialsProvider(regionType: .USEast1, identityPoolId: K.COGNITO_IDENTITY_POOL_ID, identityProviderManager:self.pool)
        
        
        AWSServiceManager.defaultServiceManager().defaultServiceConfiguration = serviceConfiguration
        
        
        // and I want to be able to do interactive authentication. So this will
        // instantiate the loginViewController when a non-authenticated user wants to
        // do something (like "GetDetails" or "GetSession")
        
        self.pool!.delegate = self
        
        self.user = self.pool?.currentUser() // if I have one this is it
```
        
Generally it makes sense to do the credentialsProvider.getIdentityId as you have indicated, why doesn't it work, it looks like it should?  Once you get the credentials, they are associated with your service configuration and you can use them.  This association was done when you did this:

`let credentialsProvider = AWSCognitoCredentialsProvider(regionType: .USEast1,
                                                            identityPoolId:"us-east-1:MY.IDENTITY_ID", identityProviderManager:pool)`

which associated you credentials provider with your pool manager (pool in this case).   And this:  

`AWSCognitoIdentityUserPool.registerCognitoIdentityUserPoolWithConfiguration(serviceConfiguration, userPoolConfiguration: userPoolConfiguration, forKey: "UserPool")
`    

That should associate the credentials provider with the user pool and service configuration.
Please take a look at the pdf, it may help clarify things.
